[Alerting][Docs] Adding query to identify long running rules to docs (#98773)
elastic/kibana: docs/user/alerting/alerting-troubleshooting.asciidoc

including ignoring server certificate validation, and providing certificate
authority data to verify servers using custom certificates.  For more details, 
see <<action-settings,Action settings>>.

[float]
[[rules-long-execution-time]]

> === Identify long running rules

Reviewer: ```suggestion
=== Identify long-running rules
```

> === Identify long running rules

Reviewer 2: We'll need to add a paragraph / section here talking about the fact that this index is a Kibana system index.  By default, no normal users should be able to read it.  Superuser can.  And presumably there is a way to allow users access to this, but I'm not quite sure what the best approach is.  Probably need to update their user to add a read priv for this index - or actually all the matching indices, and it should probably cover the versions as well, so `.kibana-event-log-*` or whatever.

Author: Added https://github.com/elastic/kibana/pull/98773/commits/5a555f400ba87dce1759e4c298a46b26b989b0f2

> The following query can help you identify rules that are taking a long time to execute and may be impacting the overall health of your deployment.

Reviewer: ```suggestion
The following query can help you identify rules that are taking a long time to execute and might impact the overall health of your deployment.
```

Query for a list of rule ids, bucketed by their execution times:

> [source,sh]

Author: Added https://github.com/elastic/kibana/pull/98773/commits/adfc19c141985ad86b658c53b33533cc94bd1e85 although I did try it out both on this command and on a command in the Elasticsearch docs and while it opened up my local Kibana, it didn't actually copy the command there :(

> [source,sh]

Reviewer 2: Interesting.  I just tried it with a cloud instance I have; used the gear tool button to point to my cloud dev console.  Clicked copy to console (or whatever).  Worked perfectly.  Except .... there's a `$` at the very beginning of the text, which I had to remove before I could execute it.  Can we get rid of that?

> [source,sh]

Author: Removed!

--------------------------------------------------
$ GET /.kibana-event-log*/_search
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "range": {
            "@timestamp": {
              "gte": "now-1d", <1>
              "lte": "now"
            }
          }
        },
        {
          "term": {
            "event.action": {

>               "value": "execute"

Author: Ooo good catch! Should I also make a note that provider can be changed if they want to find long running actions instead of rules?

>               "value": "execute"

Author: Added https://github.com/elastic/kibana/pull/98773/commits/adfc19c141985ad86b658c53b33533cc94bd1e85

            }
          }
        }
      ]
    }
  },
  "aggs": {
    "ruleIdsByExecutionDuration": {
      "histogram": {
        "field": "event.duration",

>         "interval": 10000000000 <2>

Reviewer 2: uggh; I wonder if there's anything to help here, to get these nano's scaled down to milli's.  I think I've seen suggestions that this field should be a multi-field, with a subfield for the millis, but I don't quite know how that would work.  Wonder how hard it would be to "runtime" a milli version of this field, for just this call?

Author: Good idea! Added runtime fields to the query in this commit https://github.com/elastic/kibana/pull/98773/commits/fbc650dc6b5ce5267eda7d118913fbe04d73a552

      },
      "aggs": {
        "ruleId": {
          "nested": {
            "path": "kibana.saved_objects"
          },
          "aggs": {
            "ruleId": {
              "terms": {
                "field": "kibana.saved_objects.id",
                "size": 10 <3>
              }
            }
          }
        }
      }
    }
  }
}

> --------------------------------------------------

Reviewer 2: The first time I ran this, all the executions fit in the first bucket, so ... not much value there.  Turned the interval down to get more buckets created.

Could we used a fixed bucket count here?  Like 10?  So you'd always see 10 buckets, with ES figuring out the interval.  I can't remember if that's a thing you can do in ES, but I think so?

Author: I believe date histogram allows this type of auto-interval calculation but the regular histogram aggregation does not. There is an issue for it in Elasticsearch https://github.com/elastic/elasticsearch/issues/31828

        {
          "key" : 1.0E10,
          "doc_count" : 0,
          "ruleId" : {
            "doc_count" : 0,
            "ruleId" : {
              "doc_count_error_upper_bound" : 0,
              "sum_other_doc_count" : 0,
              "buckets" : [ ]
            }
          }
        },
        {
          "key" : 2.0E10,
          "doc_count" : 0,
          "ruleId" : {
            "doc_count" : 0,
            "ruleId" : {
              "doc_count_error_upper_bound" : 0,
              "sum_other_doc_count" : 0,
              "buckets" : [ ]
            }
          }
        },
        {
          "key" : 3.0E10, <2>
          "doc_count" : 6,
          "ruleId" : {
            "doc_count" : 6,
            "ruleId" : {
              "doc_count_error_upper_bound" : 0,
              "sum_other_doc_count" : 0,
              "buckets" : [
                {
                  "key" : "41893910-6bca-11eb-9e0d-85d233e3ee35",
                  "doc_count" : 6
                }
              ]
            }
          }
        }
      ]
    }
  }
}
--------------------------------------------------
<1> Most rule execution durations fall within the first bucket (0 - 10 seconds).

> <2> A single rule with id `41893910-6bca-11eb-9e0d-85d233e3ee35` has taken between 30 and 40 seconds to execute.

Reviewer: ```suggestion
<2> A single rule with id `41893910-6bca-11eb-9e0d-85d233e3ee35` took between 30 and 40 seconds to execute.
```

> We can use the <<get-rule-api,Get Rule API>> to retrieve additional information about rules that are taking a long time to execute.

Reviewer: ```suggestion
Use the <<get-rule-api,Get Rule API>> to retrieve additional information about rules that take a long time to execute.
```

[IMPORTANT]
==============================================
By default, only users with a `superuser` role can query the {kib} event log because it is a system index. To enable additional users to execute this query, assign `read` privileges to the `.kibana-event-log*` index.
==============================================

Query for a list of rule ids, bucketed by their execution times:

[source,console]
--------------------------------------------------
$ GET /.kibana-event-log*/_search
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "range": {
            "@timestamp": {
              "gte": "now-1d", <1>
              "lte": "now"
            }
          }
        },
        {
          "term": {
            "event.action": {
              "value": "execute"
            }
          }
        },
        {
          "term": {
            "event.provider": {
              "value": "alerting" <2>
            }
          }
        }
      ]
    }
  },
  "runtime_mappings": { <3>
    "event.duration_in_seconds": {
      "type": "long",
      "script": {

>         "source": "emit(doc['event.duration'].value / 1000000000)"

Reviewer 2: awesome!

Reviewer 2: just happened to think, as I was counting the `0`'s :-), that perhaps we could use engineering notation here?  ES provides that back in it's output, I assume it would also support `1.0E9` instead of `1-with-9-zeros`.

Author: Good call! Updated.

      }
    }
  },
  "aggs": {
    "ruleIdsByExecutionDuration": {
      "histogram": {
        "field": "event.duration_in_seconds",

>         "interval": 10 <4>

Reviewer 2: So this will bucket the alerts by 10 second buckets?  I think that's what it did for me.  Everything got lumped in the first bucket again :-).  I changed this to `0.1`, which gave me buckets of that size, which worked out pretty good.  Recommend we switch to something on that order.  Hopefully, most alert executions are taking < 1 second :-)

Author: I changed the example to use a 1 second interval instead of 10 seconds.

This query returns the following:

[source,json]
--------------------------------------------------
{
  "took" : 322,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 204,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "ruleIdsByExecutionDuration" : {
      "buckets" : [
        {
          "key" : 0.0, <1>
          "doc_count" : 320,
          "ruleId" : {
            "doc_count" : 320,
            "ruleId" : {
              "doc_count_error_upper_bound" : 0,
              "sum_other_doc_count" : 0,
              "buckets" : [
                {
                  "key" : "1923ada0-a8f3-11eb-a04b-13d723cdfdc5",
                  "doc_count" : 140
                },
                {
                  "key" : "15415ecf-cdb0-4fef-950a-f824bd277fe4",
                  "doc_count" : 130
                },
                {
                  "key" : "dceeb5d0-6b41-11eb-802b-85b0c1bc8ba2",
                  "doc_count" : 50
                }
              ]
            }
          }
        },
        {

>           "key" : 1.0E10,

Reviewer 2: The keys I'm seeing in my results (where I set the histogram interval to 0.1) are 0.0, 0.1, 0.2.  I'd think the keys given an interval of 10 would be 0.0, 10.0, etc.  But these ook like the non-scripted nanosecond values instead.  Copy pasta?

Author: Updated!

The following query can help you identify rules that are taking a long time to execute and might impact the overall health of your deployment.

[IMPORTANT]
==============================================
By default, only user with a `superuser` role can query the {kib} event log because it is a system index. To enable additional users to execute this query, assign `read` privileges to the `.kibana-event-log*` index.

> ==============================================

Author: @gchaps Can you review this addition to the docs? Thank you!
